xdsl/frontend/passes/desymref.py

```diff
@@ -0,0 +1,355 @@
+#     for op in block.ops:
+#       desymref_op(op)
+#     # at this point no op contains a symref operation
+#     # in the nested region.
+# ```
+#
+# Next, we describe how desymrefication actually works, on a high level.
+# Every region either declares a symbol or uses it from the parent region.
+# Let's consider each case separately.
+#
+# Case 1: Symbol is declared in this region. This is an easy case. We already
+# know that symbol is only used in this region and no other nested region uses
+# it. Therefore running any SSA-construction algorithm on the CFG is enough.
+#
+# Case 2: Symbol is not declared in this region. In this case, we only support
+# non-CFGs at the moment (i.e. single block regions). We observe that in general
+# every symbol can be simplified within a block to have at most one fetch (start
+# of the block) at most one update (end of the block). This means that using
+# specification of an operation we can promote these symbols outside.
+# TODO: Add op promotion.
+
+
+def has_symbol(op: Operation) -> bool:
+    return isinstance(op, symref.Declare) or isinstance(
+        op, symref.Fetch) or isinstance(op, symref.Update)
+
+
+def get_symbol(op: Operation) -> str:
+    """
+    Returns a symbol attribute for an operation. If operation has no symbol
+    attributes, raises an exception.
+    """
+    if isinstance(op, symref.Declare):
+        return op.sym_name.data
+    elif isinstance(op, symref.Fetch) or isinstance(op, symref.Update):
+        return op.symbol.root_reference.data
+    else:
+        raise FrontendProgramException(
+            f"Operation '{op.name}' does not have symbol attributes.")
+
+
+def get_symbols(block: Block) -> Set[str]:
+    """Returns a set of all symbols defined/used in a basic block."""
+    symbols: Set[str] = set()
+    for op in block.ops:
+        if has_symbol(op):
+            symbols.add(get_symbol(op))
+    return symbols
+
+
+def lower_positional_bound(ops: List[Operation],
+                           op: Operation) -> Operation | None:
+    """
+    Returns an operation from `ops` which is a lower bound for `op`. If there is
+    no such operation, `None` is returned.
+    
+    Pre-condition: list `ops` is sorted for the provided numbering.
+    """
+    block = op.parent_block()
+    assert block is not None
+
+    idx = block.get_operation_index(op)
+    low_idx = -1
+    high_idx = len(ops) - 1
+
+    # Binary search to find the right operation.
+    while low_idx < high_idx:
+        mid_idx = (high_idx - low_idx + 1) // 2 + low_idx
+        user_idx = block.get_operation_index(ops[mid_idx])
+
+        if user_idx < idx:
+            low_idx = mid_idx
+        else:
+            high_idx = mid_idx - 1
+
+    if low_idx == -1:
+        return None
+    return ops[low_idx]
+
+
+@dataclass
+class Desymrefier:
+    """
+    Rewrites the program by removing all reads/writes from/to symbols and symbol
+    definitions.
+    """
+
+    rewriter: Rewriter
```

Comment:
I think you don't need this, and you could just use `Rewriter.your_function` instead.